### DISCUSSION TITLE
Automatic algorithm to avoid collision with toolhead?

### DISCUSSION THREAD

**Discussion by bitsbau** [visose/Robots#193] [Q&A] [answered]
Hi, @visose. Thank you for the amazing plugin. 

I'm trying to simulate a stacking bricks sample similar like the sample you provided. However, no matter how I changed the robot configurations' option under the target component, I couldn't get the wished pose to avoid collision. Is there someway to constraint specific joints' angle of elbow and wrist for example more precisely in the kinematics component or perhaps better solution with automatic algorithm to avoid collision with the toolhead? 

Also can you add a steps input in the simulation so that you can overview it by steps rather than time.

**Comment by visose** [comment, marked as answer]
Hi, a 6 axis robot can reach a particular target in a maximum of 8 poses, if it's reachable and the joint limits are within valid range. Within a sequence of linear motions, the robot remains in the same configuration, so there is a maximum of 1 pose. If you constrain any joints, it will not be possible to reach the target.

You can add more degrees of freedom in two ways, by adding more axes (7 axis robot, track, positioner, etc.) or by reducing the required degrees of freedom. For example, imagine a robot placing cylindrical objects on a table gripping them from the top with a vacuum gripper. If the objects are completely symmetrical around their normal axis, the robot could potentially reach the placing target with any rotation around its normal.

In any case, with this plugin you will have to explicitly define the toolpath. Try changing the relative orientation between the tool tcp and the flange, different gripping locations, different placement of the objects relative to the robot. Make sure you really need some sophisticated collision avoidance trajectory generation before you reach for it. Most of the time you don't.

**Reply by bitsbau** [reply]
Hi, @visose. Thank you for your reply. I've rechecked my file and realize that I need an algorithm to define the pose explicitly at some target milestone (especially when joint motion turns to lineal motion), in order to avoid the collision between robot itself and the end effector. 

There is always a good reachable solution among the 8 poses, however not all the targets can be settled with the same RobConf index. Like the first target I can tick only the shoulder, the next one has to be shoulder and wrist. (in joint motion because it needn't to be precise) From your example, you also did some replacement of unwinding forearm by setting unambiguous pose. What I don't understand with it was: 

the joint rotation we get from kinematics component, which describe the rotation of the robotic joints respectively. How can you these radians value to select the optimal RobConf definition? How are they related and how can I develop this algorithm? Or what problem could it result to? 

IT would be also appreciate if you could recommend me some reading materials on these basic concepts with toolpath planning. 

Thanks.

**Reply by visose** [reply]
I'm not sure if you are asking the following:
When performing a joint motion with a cartesian target, if the configuration is not explicitly set, it adopts the closest of the 8 possible configurations, and your question is, what does it mean to be the closest?

It is the closest in "joint space", this means the pose that needs to rotate the joints the least. This is done by calculating all 8 possible solutions and applying the least squares method.  So, for each, a number is calculated by first subtracting the joints values with the previous joint values, this resulting vector is then added to a number like this: j1 x j1 + j2 x j2 + j3 x j3... where jn is how much each axis has to rotate. The pose with the lowest number is chosen.
This is similar to getting the length between two points in cartesian space by subtracting them, and then getting the length of the resulting vector, but without the square root, as it is not needed to compare between distances.
